lib/backend/watchersyncer/watchersyncer.go

```diff
@@ -0,0 +1,200 @@
+	// the Syncer.
+	UpdateProcessor SyncerUpdateProcessor
+}
+
+// SyncerUpdateProcessor is used to convert a Watch update into one or more additional
+// Syncer updates.
+type SyncerUpdateProcessor interface {
+	// Process is called to process a watch update.  The processor may convert this
+	// to zero or more updates.  The processor may use these calls to maintain a local cache
+	// if required.  It is safe for the processor to send multiple duplicate adds or deletes
+	// since the WatcherSyncer maintains it's own cache and will swallow duplicates.
+	// A KVPair with a nil value indicates a delete.  A non nil value indicates an add/modified.
+	// The processor may respond with any number of adds or deletes.
+	Process(*model.KVPair) ([]*model.KVPair, error)
+
+	// SyncStarting is called when syncer is starting a full sync for the associated resource
+	// type.  That means it is first going to list current resources and then watch for any updates.
+	// If the processor maintains a private internal cache, then the cache should be cleared at
+	// this point since the cache will be re-populated from the sync.
+	SyncStarting()
+}
+
+// New creates a new multiple Watcher-backed api.Syncer.
+func New(client api.Client, resourceTypes []ResourceType, callbacks api.SyncerCallbacks) api.Syncer {
+	rs := &watcherSyncer{
+		watcherCaches: make([]*watcherCache, len(resourceTypes)),
+		results:       make(chan interface{}, 2000),
+		callbacks:     callbacks,
+	}
+	for i, r := range resourceTypes {
+		rs.watcherCaches[i] = newWatcherCache(client, r, rs.results)
+	}
+	return rs
+}
+
+// watcherSyncer implements the api.Syncer interface.
+type watcherSyncer struct {
+	status        api.SyncStatus
+	watcherCaches []*watcherCache
+	results       chan interface{}
+	numSynced     int
+	callbacks     api.SyncerCallbacks
+}
+
+func (rs *watcherSyncer) Start() {
+	log.Info("Start called")
+	go rs.run()
+}
+
+// Send a status update and store the status.
+func (rs *watcherSyncer) sendStatusUpdate(status api.SyncStatus) {
+	log.WithField("Status", status).Info("Sending status update")
+	rs.callbacks.OnStatusUpdated(status)
+	rs.status = status
+}
+
+// run implements the main syncer loop that loops forever receiving watch events and translating
+// to syncer updates.
+func (rs *watcherSyncer) run() {
+	log.Info("Running - sending status update")
+	rs.sendStatusUpdate(api.WaitForDatastore)
+	log.Info("Starting watcher caches")
+	for _, wc := range rs.watcherCaches {
+		go wc.run()
+	}
+
+	log.Info("Starting main watch processing loop")
+	var updates []api.Update
+	for {
+		// Block until there is data.
+		result := <-rs.results
+
+		// Process the data - this will append the data in subsequent calls, and action
+		// it if we hit a non-update event.
+		updates := rs.processResult(updates, result)
+
+		// Append results into the one update until we either flush the channel or we
+		// hit our fixed limit per update.
+	consolidatationloop:
```

Comment:
Is this logic worth it?

@fasaxc does Felix perform significantly better if we wait for a batch of updates over the syncer rather than sending them through individually? If so, do we have any idea what the right batch size to send is? 

@robbrockbank how did we pick `maxUpdatesToConsolidate`?

Reply:
completely random - and yup I glad you flagged this - I wanted to check with Shaun when he got back too....

Reply:
actually it wasn't random - i made it the same length as the channel capacity.  (i forgot about that)

Reply:
It may help a bit at start-of-day, when we have to push 150k updates through but at steady-state 100-1000updates/s it shouldn't make a difference.  IIRC, sending on a channel takes about half a microsecond so if you're doing 1000 sends/s it's negligible; if you're doing 100k+ then you'll start to notice and if you're doing 1m+ then it dominates.